core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -418,7 +423,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * does not fail, but succeeds with the value returned by the left or right
    * function passed to `fold`.
    */
-  final def fold[B](failure: E => B, success: A => B): ZIO[R, Nothing, B] =
+  final def fold[B](failure: E => B, success: A => B)(implicit ev: CanFail[E]): ZIO[R, Nothing, B] =
```

Comment:
Do you mean use an implicit `E =:= Nothing`? They can still use this method If E is not literal a Nothing, but only by evidence. Otherwise I'm missing what's the additional power vs. `.map(success)`, for E=Nothing the type is always `ZIO[R, Nothing, A] => ZIO[R, Nothing, B]`, i.e. `map`